### PR TITLE
feat: add Security Analyzer functionality

### PR DESCRIPTION
**Pull request by adrgs**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

This PR introduces the possibility of having a defined security analyzer watch over OpenDevin's actions and watch accordingly. It's integrated by default with [confirmation mode](https://github.com/OpenDevin/OpenDevin/pull/2774), which analyzers can make use of by auto-approving actions deemed safe. 

We have also implemented such an analyzer, using [invariant](https://github.com/invariantlabs-ai/invariant)

---

**Give a summary of what the PR does, explaining any non-trivial design decisions**

Backend
- Important actions now have the security_risk field, which is Optional
- The Security Analyzer lives in AgentSession, and can only be created at start-up if the proper flag is set
- Added a `/api/security/{path:path}` API route, for the analyzers to be able to interact with the user without having to modify the core code for each
- Also added `/api/options/security-analyzers` endpoint, which lists the agents
- Added documentation in [opendevin/security](https://github.com/invariantlabs-ai/OpenDevin/tree/security_analyzer/opendevin/security) for creating new analyzers
- The Invariant Analyzer runs in a separate Docker container, to not add any dependencies to OpenDevin

Frontend
- Added new combobox / string setting for SECURITY_ANALYZER
- Added a Lock icon in the bottom right that is visible when there's a security analyzer present
- Added `frontend/src/components/modals/security`, a place where the modules can create their own modals
---
**Other references**
Video of how the changes look like:
https://www.youtube.com/watch?v=brH7CUGyzmI

### CODE REVIEW

File: opendevin/security/invariant/parser.py
Comment (xingyaoww) [inline opendevin/security/invariant/parser.py:104]: does this function need to get update whenever "Action" definition changes? 🤔 I'm thinking if there's an easy way for us to maintain this? Or trigger a linter error of some sort when someone modified the action but forgot to change here.

**Review by xingyaoww — COMMENTED** [review]
Backend overall looks good! Excited for this -- I left a few comments.

Once we have this working well, we could experiment with interacting with user's local environment :)

File: opendevin/server/session/agent.py
Comment (adrgs) [reply opendevin/server/session/agent.py]: Good point! I've added a SecurityConfig and bundled confirmation_mode with security_analyzer in there. Also added a section for it in config.template.toml

security_analyzer is now `security_analyzer: str | None = None`

File: opendevin/security/invariant/analyzer.py
Comment (adrgs) [reply opendevin/security/invariant/analyzer.py]: Modified to `Error creating Invariant Security Analyzer container. Please check that Docker is running or disable the Security Analyzer in settings.`

File: opendevin/core/config.py
Comment (adrgs) [reply opendevin/core/config.py]: Mentioned here: https://github.com/OpenDevin/OpenDevin/pull/3058#discussion_r1696967568

File: tests/unit/test_security.py
Comment (xingyaoww) [inline tests/unit/test_security.py:1]: On a second thought, i think it will be nice to also tests whether the `Action` definition has changed in a way that will break the invariant action analyzer. ref: https://github.com/OpenDevin/OpenDevin/pull/3058#discussion_r1695511239

Comment (xingyaoww) [reply tests/unit/test_security.py:1]: Maybe like test cases specifically tests for the invariant parse_action / parser_observation function.

**Review by xingyaoww — COMMENTED** [review]
Backend LGTM now

File: tests/unit/test_security.py
Comment (adrgs) [reply tests/unit/test_security.py:1]: Included tests for parsing actions / observations in `test_security.py`

File: opendevin/events/action/browse.py
Comment (enyst) [inline opendevin/events/action/browse.py:15]:
```suggestion
    security_risk: ActionSecurityRisk | None
```

Just a minor thing, we support python 3.11+, so we don't need Optional.

File: opendevin/events/action/browse.py
Comment (enyst) [inline opendevin/events/action/browse.py:36]:
```suggestion
    security_risk: ActionSecurityRisk | None = None
```

File: opendevin/security/invariant/parser.py
Comment (enyst) [reply opendevin/security/invariant/parser.py:104]: We could make a method in Actions, `analyzer_trace`, which gets `next_id` and returns the list to append to `inv_trace`. Just thinking it may help, simply because the code will be in front of our eyes in the respective Action subclass.

Related, just wondering: these actions seem to be the current CodeAct/browser actions, e.g. it's not implemented for FileReadAction and others. Will we want all runnable actions or almost all of them? If yes, we could make sure we write definitions, by making the `analyzer_trace` method abstract in Action. This may be overkill, though.

File: opendevin/server/listen.py
Comment (enyst) [inline opendevin/server/listen.py:360]:
```suggestion
    """Get all supported security analyzers.
```

Just a minor thing, it doesn't depend on the agent, right?

File: opendevin/server/session/agent.py
Comment (enyst) [inline opendevin/server/session/agent.py:28]:
```suggestion
    security_analyzer: SecurityAnalyzer | None = None
```

File: opendevin/server/listen.py
Comment (adrgs) [reply opendevin/server/listen.py]: you're right, it doesn't

File: opendevin/security/invariant/parser.py
Comment (adrgs) [reply opendevin/security/invariant/parser.py:61]: Hello, please take a look at the new parser:

https://github.com/OpenDevin/OpenDevin/pull/3058/commits/32c08ffbba9bcfde7d913152c0779b7e9d6407ff#diff-d4f29cbf26d5f92d82b77f0ab8c0d5fa8aeeb1b67f6f4fec0e1f5c7fd04d06f1R51

Changed such that MessageAction is just a user/agent trace. NullAction and ChangeAgentStateAction don't produce traces, and all the other actions are just function calls (the name of each function is as defined in [/opendevin/core/schema/action.py](/OpenDevin/OpenDevin/blob/main/opendevin/core/schema/action.py)

Similar thing for observations.

This should help decouple the analyzer from actions, and custom policies can be written for newer actions as well. (now including FileWrite/Read)

Comment (enyst) [reply opendevin/security/invariant/parser.py:61]: Simple and clear, love it.

Thanks so much for doing this! ❤️

File: opendevin/security/README.md
Comment (xingyaoww) [inline opendevin/security/README.md:4]: Could you add a section here as a quick start to show people how to enable this feature? e.g., on the UI and on the backend?

**Review by xingyaoww — COMMENTED** [review]
Backend LGTM! If frontend also look good to folks, i'd be have to approve it.

File: opendevin/security/README.md
Comment (adrgs) [reply opendevin/security/README.md:4]: sure, added it in the latest commit:
````suggestion
To enable this feature: 
* From the web interface
    * Open Configuration (by clicking the gear icon in the bottom right)
    * Select a Security Analyzer from the dropdown
    * Save settings
    * (to disable) repeat the same steps, but click the X in the Security Analyzer dropdown
* From config.toml
```toml
[security]
# Enable confirmation mode
confirmation_mode = true
# The security analyzer to use
security_analyzer = "your-security-analyzer"
```
(to disable) remove the lines from config.toml
````

Comment (xingyaoww) [reply opendevin/security/README.md:4]: Awesome!

**Comment by xingyaoww** [comment]
Hey @adrgs, I'd be happy to merge it if you could resolve the conflict (i help did it in this PR https://github.com/invariantlabs-ai/OpenDevin/pull/1)

**Comment by adrgs** [comment]
> Hey @adrgs, I'd be happy to merge it if you could resolve the conflict (i help did it in this PR [invariantlabs-ai#1](https://github.com/invariantlabs-ai/OpenDevin/pull/1))

thanks for the help @xingyaoww ! should be up to date with main now

**Comment by neubig** [comment]
Hey @xingyaoww , could you make the final call on this and we can get it merged in? This is exciting!

**Review by xingyaoww — APPROVED** [review]
Let's get this in!

**Comment by xingyaoww** [comment]
@adrgs sorry for the late response! Could you resolve the one last merge conflict: https://github.com/invariantlabs-ai/OpenDevin/pull/2

**Comment by adrgs** [comment]
@xingyaoww added, thanks!

**Pull request merged**
